[GISel] Funnel shift combiner port from SelectionDAG ISel to GlobalISel

axelcool1234 opened the pull request:
The funnel shift combiner rule from https://github.com/llvm/llvm-project/commit/4a3708cd6b062a6afe2697cc8b39329e3f2faa25 is currently missing from GlobalISel. The following is a port of that combiner to GlobalISel.

axelcool1234 committed:
pre-commit test

axelcool1234 commented:
@mshockwave I finished that combiner port! Let me know what you think :)

davemgreen commented:
Can you update llvm/test/CodeGen/AArch64/funnel-shift.ll too?

axelcool1234 commented:
Alright I updated the AArch64 test using `update_llc_test_checks.py` (it removed the redundant left shift and it being ORed with the funnel shift in the test).

davemgreen commented on llvm/include/llvm/Target/GlobalISel/Combine.td:
```diff
@@ -1033,6 +1033,24 @@ def funnel_shift_overshift: GICombineRule<
+// Transform: fshl x, ?, y | shl x, y -> fshl x, ?, y
+def funnel_shift_or_shift_to_funnel_shift_left: GICombineRule<
+  (defs root:$root), 
+  (match (G_FSHL $out1, $x, $_, $y),
```
$_ -> $z if it gets used. $_ makes it look like an unused input, but IIUC it is just a different input variable?

davemgreen commented on llvm/test/CodeGen/AArch64/funnel-shift.ll:
```diff
@@ -674,14 +674,12 @@ define i32 @or_shl_fshl_simplify(i32 %x, i32 %y, i32 %s) {
+; CHECK-GI-NEXT:    orr w0, w10, w8
 ; CHECK-GI-NEXT:    ret
   %shy = shl i32 %y, %s
   %fun = call i32 @llvm.fshl.i32(i32 %y, i32 %x, i32 %s)
```
Do you have any idea why or_lshr_fshr_simplify below is not triggering too? An Or should be fine with commuted operands.

axelcool1234 replied:
Strange... I'm not really sure. Over the next few days I'll look over everything and see if I can reason about why this isn't working.

axelcool1234 replied:
So I've determined why it isn't triggering. If I change `(G_OR $root, $out1, $out2)),` to `(G_OR $root, $out2, $out1)),`, it matches the `or_lshr_fshr_simplify` test case:

```
  %shy = lshr i32 %y, %s
  %fun = call i32 @llvm.fshr.i32(i32 %x, i32 %y, i32 %s)
  %or = or i32 %shy, %fun
  ret i32 %or
```

I can also change `%or = or i32 %shy, %fun` to `%or = or i32 %fun, %shy` and make it trigger too. I'm new to the LLVM project but I had the assumption that `G_OR` in tablegen didn't care for the ordering of operands, but it seems like that isn't the case here. Perhaps I'm missing something? @davemgreen @mshockwave

axelcool1234 replied:
> An Or should be fine with commuted operands.

I just realized this is what you were referring to. I've personally confirmed it does seem to be due to the OR statement at least. I'm not sure why this is happening...

davemgreen reviewed (commented):
Thanks. The combine looks good - nice and clean - other than the points mentioned below.

We would not usually need to add combines that the midend can reliably perform unless they come up during lowering. I am not sure if that is expected in this case or not, from the legalization of higher sized types or from vectors.

arsenm commented on llvm/include/llvm/Target/GlobalISel/Combine.td:
```diff
@@ -1033,6 +1033,24 @@ def funnel_shift_overshift: GICombineRule<
+>;
+
+// Transform: fshr ?, x, y | srl x, y -> fshr ?, x, y
+def funnel_shift_or_shift_to_funnel_shift_right: GICombineRule<
+  (defs root:$root), 
+  (match (G_FSHR $out1, $_, $x, $y),
+         (G_LSHR $out2, $x, $y),
+         (G_OR $root, $out1, $out2)),
```
Missing hasOneUse checks?

axelcool1234 replied:
> Missing hasOneUse checks?

The original SelectionDAG ISel didn't have a check for this. I do see how it'd be problematic for the `G_OR` and `G_LSHR` statements to be removed if later code relied on them. How did the SelectionDAG ISel combiner not case this problem?

I also couldn't manage to find any Tablegen code utilizing `hasOneUse`.  If you have any resources or examples, I'd be immensely grateful!

arsenm replied:
Proper tablegen support for hasOneUse is a missing feature, other places have to drop into the match function in c++

arsenm replied:
> The original SelectionDAG ISel combiner didn't have a check for this. 

Common combiner bug, if you need to emit those instructions it's generally bad

arsenm commented on llvm/test/CodeGen/RISCV/GlobalISel/shift.ll:
```diff
@@ -105,3 +105,55 @@ define i16 @test_shl_i48_2(i48 %x, i48 %y) {
+; RV64-NEXT:    or a0, a0, a1
+; RV64-NEXT:    ret
+
+    %fshl = call i32 @llvm.fshl.i32(i32 %x, i32 %_, i32 %y)
+    %shl = shl i32 %x, %y
+    %or = or i32 %fshl, %shl
+    %trunc = trunc i32 %or to i16
+    ret i16 %trunc
```
Over-indented

arsenm commented on llvm/test/CodeGen/RISCV/GlobalISel/shift.ll:
```diff
@@ -105,3 +105,55 @@ define i16 @test_shl_i48_2(i48 %x, i48 %y) {
+    %or = or i32 %fshr, %lshr
+    %trunc = trunc i32 %or to i16
+    ret i16 %trunc
+}
```
Vector version of tests? Also negative multi-use tests. Ideally would share the existing DAG combine tests here instead of creating new copies

axelcool1234 replied:
> Ideally would share the existing DAG combine tests here instead of creating new copies

The issue I ran into has to do with the i48 type leading to a lot of code being generated to emulate `@llvm.fshl.i48` since it isn't natively supported. The resulting expected test output would be really large. Should I still go for it? 

> Also negative multi-use tests.

Is this referring to the fact that I'm truncating or something else? I apologize, I'm still getting a hang to contributing here (and thanks for reviewing my patch!)

> Vector version of tests?

Is there a common vector size that many contributors use for vector versions of tests? I'll make sure to implement that :)

arsenm replied:
> The issue I ran into has to do with the i48 type leading to a lot of code being generated to emulate @llvm.fshl.i48 since it isn't natively supported. The resulting expected test output would be really large. Should I still go for it?

That's going to be true in either case. As long as it actually compiles, it's not a problem and gives a point of reference for whenever the code improves for illegal types.

Ideally would share the existing DAG combine tests here instead of creating new copies
The issue I ran into has to do with the i48 type leading to a lot of code being generated to emulate @llvm.fshl.i48 since it isn't natively supported. The resulting expected test output would be really large. Should I still go for it?

Also negative multi-use tests.
Is this referring to the fact that I'm truncating or something else? I apologize, I'm still getting a hang to contributing here (and thanks for reviewing my patch!)

> Is there a common vector size that many contributors use for vector versions of tests? I'll make sure to implement that :)

No, whatever is relevant for the target (you also wouldn't need to think too hard about that if reusing the original test, which I hope covers some vectors)

axelcool1234 replied:
> That's going to be true in either case. As long as it actually compiles, it's not a problem and gives a point of reference for whenever the code improves for illegal types.

Before I go forward and introduce i48 funnel shift calls to the original tests (in order to also test for multi-use and if it correctly ensures the intermediate instructions are still emitted), I want to further clarify what I meant. The following is a test that uses an i32 fshl and another test that uses an i48 fshl. The difference in generated code (via `update_llc_test_checks`) is quite large; I assume that isn't much of a concern here? Once I introduce these calls to the original tests their generated code checks will also explode in size.  

```
define i32 @test_fshl_i32(i32 %x, i32 %y) {
; RV32-LABEL: test_fshl_i32:
; RV32:       # %bb.0:
; RV32-NEXT:    not a2, a1
; RV32-NEXT:    sll a0, a0, a1
; RV32-NEXT:    li a1, 8
; RV32-NEXT:    srl a1, a1, a2
; RV32-NEXT:    or a0, a0, a1
; RV32-NEXT:    ret
;
; RV64-LABEL: test_fshl_i32:
; RV64:       # %bb.0:
; RV64-NEXT:    not a2, a1
; RV64-NEXT:    sllw a0, a0, a1
; RV64-NEXT:    li a1, 8
; RV64-NEXT:    srlw a1, a1, a2
; RV64-NEXT:    or a0, a0, a1
; RV64-NEXT:    ret
  %fshl = call i32 @llvm.fshl.i32(i32 %x, i32 16, i32 %y)
  ret i32 %fshl
}

define i48 @test_fshl_i48(i48 %x, i48 %y) {
; RV32-LABEL: test_fshl_i48:
; RV32:       # %bb.0:
; RV32-NEXT:    addi sp, sp, -16
; RV32-NEXT:    .cfi_def_cfa_offset 16
; RV32-NEXT:    sw ra, 12(sp) # 4-byte Folded Spill
; RV32-NEXT:    sw s0, 8(sp) # 4-byte Folded Spill
; RV32-NEXT:    sw s1, 4(sp) # 4-byte Folded Spill
; RV32-NEXT:    sw s2, 0(sp) # 4-byte Folded Spill
; RV32-NEXT:    .cfi_offset ra, -4
; RV32-NEXT:    .cfi_offset s0, -8
; RV32-NEXT:    .cfi_offset s1, -12
; RV32-NEXT:    .cfi_offset s2, -16
; RV32-NEXT:    mv s1, a0
; RV32-NEXT:    mv s0, a1
; RV32-NEXT:    mv a0, a2
; RV32-NEXT:    li s2, 47
; RV32-NEXT:    slli a1, a3, 16
; RV32-NEXT:    srli a1, a1, 16
; RV32-NEXT:    li a2, 48
; RV32-NEXT:    li a3, 0
; RV32-NEXT:    call __umoddi3
; RV32-NEXT:    li a2, 32
; RV32-NEXT:    bltu a0, a2, .LBB2_2
; RV32-NEXT:  # %bb.1:
; RV32-NEXT:    li a1, 0
; RV32-NEXT:    sll a4, s1, a0
; RV32-NEXT:    sub a3, s2, a0
; RV32-NEXT:    bnez a0, .LBB2_3
; RV32-NEXT:    j .LBB2_4
; RV32-NEXT:  .LBB2_2:
; RV32-NEXT:    sll a1, s1, a0
; RV32-NEXT:    neg a3, a0
; RV32-NEXT:    srl a3, s1, a3
; RV32-NEXT:    sll a4, s0, a0
; RV32-NEXT:    or a4, a3, a4
; RV32-NEXT:    sub a3, s2, a0
; RV32-NEXT:    beqz a0, .LBB2_4
; RV32-NEXT:  .LBB2_3:
; RV32-NEXT:    mv s0, a4
; RV32-NEXT:  .LBB2_4:
; RV32-NEXT:    li a0, 8
; RV32-NEXT:    bltu a3, a2, .LBB2_6
; RV32-NEXT:  # %bb.5:
; RV32-NEXT:    li a2, 0
; RV32-NEXT:    bnez a3, .LBB2_7
; RV32-NEXT:    j .LBB2_8
; RV32-NEXT:  .LBB2_6:
; RV32-NEXT:    srl a2, a0, a3
; RV32-NEXT:    beqz a3, .LBB2_8
; RV32-NEXT:  .LBB2_7:
; RV32-NEXT:    mv a0, a2
; RV32-NEXT:  .LBB2_8:
; RV32-NEXT:    or a0, a1, a0
; RV32-NEXT:    mv a1, s0
; RV32-NEXT:    lw ra, 12(sp) # 4-byte Folded Reload
; RV32-NEXT:    lw s0, 8(sp) # 4-byte Folded Reload
; RV32-NEXT:    lw s1, 4(sp) # 4-byte Folded Reload
; RV32-NEXT:    lw s2, 0(sp) # 4-byte Folded Reload
; RV32-NEXT:    .cfi_restore ra
; RV32-NEXT:    .cfi_restore s0
; RV32-NEXT:    .cfi_restore s1
; RV32-NEXT:    .cfi_restore s2
; RV32-NEXT:    addi sp, sp, 16
; RV32-NEXT:    .cfi_def_cfa_offset 0
; RV32-NEXT:    ret
;
; RV64-LABEL: test_fshl_i48:
; RV64:       # %bb.0:
; RV64-NEXT:    addi sp, sp, -32
; RV64-NEXT:    .cfi_def_cfa_offset 32
; RV64-NEXT:    sd ra, 24(sp) # 8-byte Folded Spill
; RV64-NEXT:    sd s0, 16(sp) # 8-byte Folded Spill
; RV64-NEXT:    sd s1, 8(sp) # 8-byte Folded Spill
; RV64-NEXT:    .cfi_offset ra, -8
; RV64-NEXT:    .cfi_offset s0, -16
; RV64-NEXT:    .cfi_offset s1, -24
; RV64-NEXT:    mv s0, a0
; RV64-NEXT:    li s1, 47
; RV64-NEXT:    slli a0, a1, 16
; RV64-NEXT:    srli a0, a0, 16
; RV64-NEXT:    li a1, 48
; RV64-NEXT:    call __umoddi3
; RV64-NEXT:    subw s1, s1, a0
; RV64-NEXT:    sll a0, s0, a0
; RV64-NEXT:    li a1, 8
; RV64-NEXT:    srl a1, a1, s1
; RV64-NEXT:    or a0, a0, a1
; RV64-NEXT:    ld ra, 24(sp) # 8-byte Folded Reload
; RV64-NEXT:    ld s0, 16(sp) # 8-byte Folded Reload
; RV64-NEXT:    ld s1, 8(sp) # 8-byte Folded Reload
; RV64-NEXT:    .cfi_restore ra
; RV64-NEXT:    .cfi_restore s0
; RV64-NEXT:    .cfi_restore s1
; RV64-NEXT:    addi sp, sp, 32
; RV64-NEXT:    .cfi_def_cfa_offset 0
; RV64-NEXT:    ret
  %fshl = call i48 @llvm.fshl.i48(i48 %x, i48 16, i48 %y)
  ret i48 %fshl
}
```

> you also wouldn't need to think too hard about that if reusing the original test, which I hope covers some vectors

There aren't any vector tests here, but I'll gladly add some myself!

mshockwave replied:
Personally I don't think you have to test i48, which is an illegal type for RISC-V. Because the majority of the instructions you saw in i48's case are generated by the legalizer, which is unrelated to your patch here. So for the purpose of testing your pattern, it'll be more clear and succinct to only include legal types (i.e. i32 and i64)

> Also negative multi-use tests.

I think this related to a check that is potentially missing from your pattern now: ideally, the pattern would replace three instructions
```
(G_FSHR $out1, $z, $x, $y),
(G_LSHR $out2, $x, $y),
(G_OR $root, $out1, $out2)
```
into just one:
```
(G_FSHR $root, $z, $x, $y)
```
But if `$out2` or `$out1` have more than one user, than you cannot remove the intermediate `G_LSHR` or `G_FSHR`, respectively. In which case you have to emit the following sequence to ensure the correctness
```
(G_FSHR $out1, $z, $x, $y),
(G_LSHR $out2, $x, $y),
(G_FSHR $root, $z, $x, $y)
```
But as you might already notice, this "new" sequence of 3 instructions is nowhere better than the original sequence, which also have 3 instructions (in most architectures, G_FSHR will not be faster than G_OR). So in the case of `$out2` or `$out1` having more than one user -- it makes more sense to just _not_ do any optimization. I don't think your current pattern would conditionally trigger based on the number of users of intermediate values. You can checkout how other patterns handle it. Guarding an optimization pattern with checks on whether the intermediate values have one use is a pretty common thing among many optimizations. If you want to learn more, you can search `hasOneUse` or `m_OneUse` in DAGCombiner to see some examples.

Finally, a negative test is a test that makes sure your optimization does _not_ trigger under certain conditions. In your case, it would be an input sequence with G_FSHR or G_LSHR that have more than one users, and the CHECK lines should pledge that they won't be optimized.

axelcool1234 replied:
I apologize for my two week delay - I got severely sick in that time.

I have changed the combiner to check for `hasOneUse` on `$out2` uses and I now replace uses of `$root` with `$out1` if `$out2` only has one use. Additionally, I have changed the original tests `test_lshr_i48` and `test_shl_i48` to `test_lshr_i32` and `test_shl_i32`. They serve as negative multi-use tests for the combiner now. I have yet to write vector tests - that's the next thing on my list.

mshockwave commented on llvm/test/CodeGen/RISCV/GlobalISel/shift.ll:
```diff
@@ -105,3 +105,53 @@ define i16 @test_shl_i48_2(i48 %x, i48 %y) {
   ret i16 %trunc
 }
+
+define i16 @test_fshl_i32(i32 %x, i32 %_, i32 %y) {
```
could you create two commits for this PR: the first one contains only the new test functions you added (e.g. test_fshl_i32 here) but with CHECK lines showing the unoptimized assembly code. And the second commit being your changes, in which the CHECK lines are updated with optimized code.
With such, it's easier to see the changes made by your patch by comparing the differences of the CHECK lines between the two commits. The first commit is also known as "pre-commit test"

axelcool1234 replied:
> Please refrain from force push unless it's necessary: https://www.llvm.org/docs/GitHub.html#rebasing-pull-requests-and-force-pushes. Please append new commits instead, as they're all gonna be squash into one at the end.

Woops! I apologize, I can't believe I forgot this.

> could you create two commits for this PR: the first one contains only the new test functions you added (e.g. test_fshl_i32 here) but with CHECK lines showing the unoptimized assembly code. And the second commit being your changes, in which the CHECK lines are updated with optimized code. With such, it's easier to see the changes made by your patch by comparing the differences of the CHECK lines between the two commits. The first commit is also known as "pre-commit test"

I have pushed two commits following this style, although I'm probably going to have to do it again once I introduce vector tests and (hopefully) figure out what's causing the AArch64 fshr test to not trigger the combiner.

mshockwave replied:
(not this patch) next time, you could make the first commit the pre-commit test, followed by your actual patch, such that you don't have to comment & un-comment like you did here.

mshockwave reviewed (commented):
Please refrain from force push unless it's necessary: https://www.llvm.org/docs/GitHub.html#rebasing-pull-requests-and-force-pushes.
Please append new commits instead, as they're all gonna be squash into one at the end.

arsenm commented on llvm/include/llvm/Target/GlobalISel/Combine.td:
```diff
@@ -1033,6 +1033,26 @@ def funnel_shift_overshift: GICombineRule<
+  (match (G_FSHL $out1, $x, $z, $y),
+         (G_SHL $out2, $x, $y),
+         (G_OR $root, $out1, $out2),
+         [{ return MRI.hasOneUse(${out2}.getReg()); }]),
```
```suggestion
         [{ return MRI.hasOneNonDBGUse(${out2}.getReg()); }]),
```

arsenm commented on llvm/include/llvm/Target/GlobalISel/Combine.td:
```diff
@@ -1033,6 +1033,26 @@ def funnel_shift_overshift: GICombineRule<
+  (match (G_FSHR $out1, $z, $x, $y),
+         (G_LSHR $out2, $x, $y),
+         (G_OR $root, $out1, $out2),
+         [{ return MRI.hasOneUse(${out2}.getReg()); }]),
```
```suggestion
         [{ return MRI.hasOneNonDBGUse(${out2}.getReg()); }]),
```

jayfoad commented:
I don't see why the hasOneUse checks are needed, and the DAG version that you copied does not have them.

axelcool1234 commented:
> I don't see why the hasOneUse checks are needed, and the DAG version that you copied does not have them.

I was told by @arsenm that it was a common combiner bug to lack the check. However, I do wonder:
@mshockwave You gave such a good explanation here https://github.com/llvm/llvm-project/pull/135132#discussion_r2051788654 about what's going on but I'm wondering if it's slightly incorrect now that we're applying a `GIReplaceReg`: In the combiner's current form, `$root` (which is the `OR` instruction) is replaced with the original `FSHR` instruction. If there are multiple uses of the `LSHR` instruction, would a `hasOneUse` check for that actually change anything? When `LSHR` has only one use, the replacement of the `OR` instruction leads to it having zero uses and not being emitted. If the `LSHR` has several uses, I suspect it'll be emitted as usual whether or not we have this `hasOneUse` check here since the combiner isn't removing it anymore (since it's simply replacing the `OR` instruction with the original `FSHR` instruction), right? Is my assessment of this correct?

I think my confusion stems from what happens when we do `(apply (G_FSHR $root, $z, $x, $y))` vs `(apply (GIReplaceReg $root, $out1))`. This is how I believe how it works:
- The first `apply` I mentioned removes all of the instructions that are matched with the resulting instruction in the `apply`
- The second `apply` I mentioned replaces the specified instruction

If that's correct, then I think @jayfoad might be right: if the `LSHR` instruction lacks other users, it'll never be emitted anyways (while it will if it does have users). Thus, we don't need the `hasOneUse` check.

jayfoad commented:
> I think my confusion stems from what happens when we do `(apply (G_FSHR $root, $z, $x, $y))` vs `(apply (GIReplaceReg $root, $out1))`. This is how I believe it works:
> 
> * The first `apply` I mentioned removes all of the instructions that are matched with the resulting instruction in the `apply`
> * The second `apply` I mentioned replaces _only_ the specified instruction, not everything that's matched

The combine doesn't really remove the original instructions, it just replaces all uses of the original root instruction. Then it's up to the normal dead code elimination to remove the original instructions _if_ they are unused. The point of a hasOneUse check is to make sure that one of the sub-instructions in the pattern _will_ be unused, because it doesn't have any other uses outside of the pattern we matched.

Typically for a cmobine that creates a new instruction, you would use hasOneUse checks, because you don't want to create a new instruction unless you can remove one or more of the existing instructions. But this combine does not create any new instructions, so I don't think there's any particular need for tha hasOneUse check.

axelcool1234 commented:
Now that the `hasOneUse` checks were determined to not be required, the original `lshr` and `shl` tests, which were converted into negative multi-use tests for funnel shifts, are no longer negative multi-use tests (they now trigger despite multiple uses of the intermediate instructions). Because of this, perhaps it'd be best to change these tests back to what they were before? I still have two new tests I've introduced that test `fshl` and `fshr`. 

Additionally, this file lacks vector versions of any of the shift tests. I'll look into introducing them.

axelcool1234 commented on llvm/include/llvm/Target/GlobalISel/Combine.td:
```diff
@@ -1034,20 +1034,26 @@ def funnel_shift_overshift: GICombineRule<
-         (G_OR $root, $out1, $out2)),
+  (match (funnel_shift_or_shift_to_funnel_shift_left_frags $root, $out1, $out2)),
   (apply (GIReplaceReg $root, $out1))
 >;
```
This now generates a pattern for both ordering of the `G_OR` instructions' operands. My combines now trigger the AArch test that hasn't been triggering this entire time (which was due to the operands being flipped in the or instruction).

Perhaps there's a better way to write this. Perhaps this is a workaround for something that shouldn't even be a problem (I haven't found any code handling commutativity like this). Please let me know of any improvements or changes I should make to this.

axelcool1234 replied:
@davemgreen You pointed out the fact the AArch test wasn't triggering (thanks, by the way). Do you think this is an appropriate fix for this?

davemgreen replied:
It would ideally be handled automatically by the gisel tablegen combining infrastructure (which should know that G_OR is commutative).  That sounds like a separate issue though and this looks OK to me in the meantime. It might be worth adding a comment that the lack of commutativity is why the extra pattern has been added. I was kind of surprised that this wasn't handled already.

axelcool1234 commented:
> Additionally, this file lacks vector versions of any of the shift tests. I'll look into introducing them.

Does RISCV support vector types? Every time I try to write a trivial function test I get an error stating that RISCV is unable to "lower" the arguments.

mshockwave commented:
> > Additionally, this file lacks vector versions of any of the shift tests. I'll look into introducing them.
> 
> According to the docs, RISC-V supports vectory types: https://llvm.org/docs/RISCV/RISCVVectorExtension.html. However, every time I try to write a trivial function test with a vector I get an error stating that RISC-V is unable to "lower" the arguments. I think the reason I can't use them has to do with the test running a version of RISC-V without the vector extension enabled. What should I change the `RUN` commands to in order to test `fshl` and `fshr` with vectors?

To enable vectors, you can `llc -mattr=+v ...`

axelcool1234 commented:
@mshockwave So I've been exposed to scalable vectors for the first time - I've been exploring how they work in RISC-V (its very interesting). Please tell me if I'm mistaken, but it seems GlobalISel only supports scalable vectors (and not fixed vectors) for arguments to functions. Despite this, it seems most scalable vector instructions are not supported in GlobalISel. All of my fiddling around with LLVM IR is getting me `LLVM ERROR: unable to translate instruction: call` and `LLVM ERROR: unable to lower arguments:` errors. Investigation online leads me to believe that my observations are correct: LLVM lacks [a lot of scalable vector instruction support](https://github.com/sampsyo/cs6120/issues/410). This [talk](https://llvm.org/devmtg/2023-05/slides/Tutorial-May11/01-Bradbury-GlobalISelTutorial.pdf) from 2023 also says on one slide that GlobalISel lacks scalable vector support.

If this is all true, how should I construct these vector tests? Since I can't seem to pass fixed length vectors as arguments, I'm forced to write them in the function body (but this leads to the compiler folding it all down since it statically knows the content of said fixed vectors). Unless there's a better way, perhaps I should look into loading the vector from a pointer or something? 

Admittedly, I feel I haven't done enough digging and deeper reading into this. If you have any resources I should really sink my teeth into, I will do so. Thanks again!

axelcool1234 commented:
@mshockwave it's been a while (I personally got busy), but I'm still interested in getting this merged. It's basically ready (beyond needing to rebase to the most recent commits of course), I just need some input on the vector tests another reviewer mentioned. The original SelectionDAG ISel version of this didn't have any, and the RISCV backend of GlobalISel is making it tricky to write vector tests. Do I even need them here? If so, can you read my previous comment and let me know how I should proceed? I want to thank you again for taking the time to help me get me involved in the LLVM codebase.

axelcool1234 commented:
> > @mshockwave it's been a while (I personally got busy), but I'm still interested in getting this merged. It's basically ready (beyond needing to rebase to the most recent commits of course), I just need some input on the vector tests another reviewer mentioned. The original SelectionDAG ISel version of this didn't have any, and the RISCV backend of GlobalISel is making it tricky to write vector tests. Do I even need them here? If so, can you read my previous comment and let me know how I should proceed? I want to thank you again for taking the time to help me get involved in the LLVM codebase.
> 
> Oh nice! I'm glad you're still keen to work on this. Sorry I saw your previous comment but I was pretty occupied back then.
> GISel's support on scalable vectors is indeed not as robust as its fixed vector counterpart. And RISCV is only able to translate a small set of instructions with scalable vectors. Switching to fixed vectors does help a lot either, because there appears to be some lacking legalizations on instructions with fixed vectors in RISCV's GISel backend.
> 
> So personally I think you could go ahead with only the scalar tests. Please rebase your PR and I'll give another round of review tomorrow morning.
> 
> Thanks again for pushing this forward!

Awesome, I'll make sure to rebase tomorrow. I'd like to email you too about continuing my contributions to LLVM (i.e. what I should explore/do next) under your guidance - I know there's been a 3 month gap but I'd love to keep going if you're up for it.

axelcool1234 commented on llvm/test/CodeGen/RISCV/GlobalISel/shift.ll:
```diff
@@ -105,3 +105,55 @@ define i16 @test_shl_i48_2(i48 %x, i48 %y) {
+  %fshl = call i32 @llvm.fshl.i32(i32 %x, i32 %_, i32 %y)
+  %shl = shl i32 %x, %y
+  %or = or i32 %fshl, %shl
+  %and = and i32 %or, %fshl
+  %trunc = trunc i32 %and to i16
```
If I recall correctly, I wrote the `%and` instruction because the transformation would make it `and` the same instruction to itself, making it redundant, thus the correct output would not have the `and` instruction outputted. If you want, I can copy these tests and remove that bit so it's _just_ the instructions that should be transformed.

The `trunc` instruction follows the same pattern as the other tests in the same file, and isn't really related. I can remove this too.

Should I go ahead and do these two things?

axelcool1234 replied:
Honestly, the combine, as written in `Combine.td `, makes it clear that `%or` is replaced (via `GIReplaceReg`), so I could just remove this check entirely too.

axelcool1234 replied:
@mshockwave I have made a commit with these redundant instructions removed. I also ran `update_llc_test_checks.py` just to make sure, and it doesn't change the output. 

If this looks good, I can squash all of my commits into a pre-commit test and a commit with my changes, as you mentioned [here](https://github.com/llvm/llvm-project/pull/135132#discussion_r2051782062).

mshockwave replied:
> About the one-use check: Since I'm calling GIReplaceReg, the original OR instruction (which is the root instruction of this pattern), is replaced with the funnel shift instruction, but doesn't remove the shift instruction, which means a one-use check in the TableGen code is not required. I realized this when talking with another reviewer on this PR. This https://github.com/llvm/llvm-project/pull/135132#issuecomment-2854306448 says it more succinctly than I could.

Yeah you're right, if you're replacing the result with one of its operands, rather than creating a new instruction, than it'll always be a win, because at least one `G_OR` will be removed. 

> As for the negative tests let me know if the following commit is fine, which includes the tests I had removed from this comment chain, along with the normal test cases without the additional and instruction.
> Should I add negative test cases for AArch64 (since we have normal test cases for them already)?

Originally the negative test case was stem from the one use check: creating a negative test that has more than one use to make sure the optimization doesn't kick in. But since there is no need for such check, I think your current set of tests is sufficient.

mshockwave commented on llvm/test/CodeGen/RISCV/GlobalISel/shift.ll:
```diff
@@ -105,3 +105,51 @@ define i16 @test_shl_i48_2(i48 %x, i48 %y) {
   ret i16 %trunc
 }
+
+define i32 @test_fshl_i32(i32 %x, i32 %_, i32 %y) {
```
you could add `signext` to each of these arguments so that all the `RV32` and `RV64` checks could be merged (the UTC script will do the merge for you)

axelcool1234 replied:
Alright, I added `signext` to the arguments and return type but it didn't seem to allow the UTC script to merge the checks. Note, the file says to run UTC with version 5, if that makes any difference.

mshockwave replied:
Oops, I must have read too fast and thought those widening instructions (e.g. sllw) were for widening arguments. Nvm then, you could remove those signext if you want.

mshockwave reviewed (approved):
LGTM, cheers!

axelcool1234 committed:
[GISel] Funnel shift combiner port from SelectionDAG ISel to GlobalISel

The funnel shift combiner rule from 4a3708c is currently missing from
GlobalISel. The following is a port of that combiner to GlobalISel.

axelcool1234 commented:
@mshockwave I have made a pre-commit test and a post-commit with the PR's title and description. Let me know if this is correct! And if so, then this can be merged :)

mshockwave commented:
> @mshockwave I have made a pre-commit test and a post-commit with the PR's title and description. Let me know if this is correct! And if so, then this can be merged :)

It's merged now. For future reference: you don't have to squash commits manually because Github would do that for you upon merging ("Squash and Merge" is the only option for LLVM at this moment). The pre-commit test system is also just for reviewing purposes (reviewers could navigate between different commits to see the difference), you don't really need to split it out for merging as they'll all gonna squashed into one at the end.
